frontend/model/contracts/group.js

```diff
@@ -312,32 +312,138 @@ sbp('chelonia/defineContract', {
+            distributionDate, distributionPeriodLength
+          ).toISOString()
+          if (recentDate >= extrapolatedDistributionDate) {
+            return dateToPeriodStamp(extrapolatedDistributionDate)
+          }
+          return dateToPeriodStamp(distributionDate)
+        }
+        // For dates before the distribution date, we check the stored data first,
+        // as simply substracting the default period length won't give the previous period stamp
+        // if the distribution date was updated during that period.
+        if (!sortedPeriodKeys.length) {
+          // Looks like we're in the waiting period but haven't stored it yet.
+          const waitingPeriodStamp = dateToPeriodStamp(
+            addTimeToDate(distributionDate, -distributionPeriodLength)
+          )
+          return recentDate >= dateFromPeriodStamp(waitingPeriodStamp).toISOString() ? waitingPeriodStamp : undefined
+        }
+        const oldestKnownStamp = sortedPeriodKeys[0]
+        const latestKnownStamp = sortedPeriodKeys[sortedPeriodKeys.length - 1]
+        if (recentDate < oldestKnownStamp) return
+        if (recentDate >= latestKnownStamp) {
+          const extrapolatedPeriodStamp = dateToPeriodStamp(
+            addTimeToDate(dateFromPeriodStamp(latestKnownStamp), distributionPeriodLength)
+          )
+          if (recentDate >= extrapolatedPeriodStamp) return extrapolatedPeriodStamp
+          return latestKnownStamp
+        }
+        for (let i = 1; i < sortedPeriodKeys.length; i++) {
+          if (recentDate < sortedPeriodKeys) return sortedPeriodKeys[i - 1]
+        }
+        // This should not happen
       }
     },
+    // May return 'undefined', in which case the caller should check archived data.
     periodBeforePeriod (state, getters) {
       return (periodStamp: string) => {
-        const len = getters.groupSettings.distributionPeriodLength
-        return dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(periodStamp), -len))
+        if (!isPeriodStamp(periodStamp)) throw new TypeError('must be periodStamp')
+        const { distributionDate, distributionPeriodLength } = getters.groupSettings
+        if (!distributionDate) return
+        // This is not always the current period stamp.
+        const distributionDateStamp = dateToPeriodStamp(distributionDate)
+        const sortedPeriodKeys = getters.groupSortedPeriodKeys
+        const latestKnownStamp = sortedPeriodKeys[sortedPeriodKeys.length - 1]
+        // Maybe the distribution date has just been updated,
+        // but the corresponding period has not started or is not stored yet.
+        // So we try to use the distribution date first.
+        if (periodStamp > distributionDateStamp) {
+          // Only extrapolate one period length in the future.
+          const onePeriodLenghtAhead = addTimeToDate(distributionDate, distributionPeriodLength)
+          if (periodStamp === dateToPeriodStamp(onePeriodLenghtAhead)) return distributionDate
```

Comment:
> I think we don't need to make extrapolatedPeriodstamp here
I'd be slightly reluctant to no longer return `undefined` for dates that are more than a full period length ahead of `latestKnownStamp`, but if a majority thinks otherwise then OK 👍